internal/backend/user_state.go

```diff
@@ -26,6 +26,44 @@ func (user *user) newState(metadataID remote.ConnMetadataID) (*State, error) {
 	return state, nil
 }
 
+// deleteUnusedMessagesMarkedDeleted Will delete all messages that have been marked for deletion that have are not in
+// use by any other states associated with this user.
+func (user *user) deleteUnusedMessagesMarkedDeleted(ctx context.Context, state *State) error {
+	// Don't run this code if the context has been cancelled (e.g: Server shutdown).
+	if ctx.Err() != nil {
+		return nil
+	}
+
+	return user.tx(ctx, func(tx *ent.Tx) error {
+		messageIDsMarkForDelete, err := txGetMessageIDsMarkedDeleted(ctx, tx)
+		if err != nil {
+			return err
+		}
+
+		user.statesLock.Lock()
+		defer user.statesLock.Unlock()
+
+		messageIDsToDelete := xslices.Filter(messageIDsMarkForDelete, func(id string) bool {
+			for _, st := range user.states {
+				if st == state || st.snap == nil {
+					continue
+				}
+				if st.snap.hasMessage(id) {
+					return false
+				}
+			}
+
+			return true
+		})
```

Comment:
We already have `user.hasStateInMailboxWithMessage`, which returns whether an active state exists which is selected in the given mailbox and contains the given message. It would be useful to introduce a similar method `user.hasStateWithMessage`, which would return whether an active state exists which contains the given message (regardless of which mailbox is selected):

```go
func (user *user) hasStateWithMessage(messageID string) bool {
	user.statesLock.RLock()
	defer user.statesLock.RUnlock()

	return xslices.CountFunc(maps.Values(user.states), func(state *State) bool {
		return state.snap != nil && state.snap.hasMessage(messageID)
	}) > 0
}
```

That would allow you to simplify this block:

```suggestion
		messageIDsToDelete := xslices.Filter(messageIDsMarkForDelete, func(messageID string) bool {
			return user.hasStateWithMessage(messageID)
		})
```

Reply:
While your example looks shorter, we have to lock/unlock countless of times. This version, while less condensed, only acquires the locks once.

Reply:
While the code changes you propose do lead to smaller code. We are forced to lock/unlock for every message id. This approach while less elegant, leads to less locking.

Reply:
I'd prefer readability and maintainability here over raw performance.

Reply:
It's usually bad design and a source of contention. Right now we are not deleting many messages, but if you start deleting 100s of message, this will lead to contention and can affect other areas of the code that also interact with this lock. 

I can't comment on whether this is with Go, but most RWLock implementations from the OS favor either reading or writing, this could also lead to starvation in that case.

Since we have to access all the states, it's best to lock it once and perform our query. 

In terms of readability, I can move the filter section into a separate function, as I now realized the lock scope can be reduced.